src/view/mod.rs

```diff
@@ -10,39 +14,220 @@ use tui::{
-            AppMode::Command(CommandMode::Read) => "Command mode: Opening a file",
-            AppMode::Edit => "Edit mode",
-        };
-
+        let tagged_text = app.get_tagged_text(); // Get a copy of the text to be rendered
+                                                 // For now let's not do anything fancy formatting
+        let text: Vec<_> = highlighter::highlight_tagged_text(&tagged_text, &self.tag_to_func);
         self.terminal.draw(|mut f| {
             let size = f.size();
             let block = Paragraph::new(text.iter())
-                .block(Block::default().title(title).borders(Borders::ALL))
+                .block(Block::default().title("Paragraph").borders(Borders::ALL))
                 .style(Style::default().fg(Color::White).bg(Color::Black))
                 .alignment(Alignment::Left)
                 .wrap(true);
             f.render_widget(block, size);
         })?;
+        self.terminal.hide_cursor()?;
+
         Ok(())
     }
 }
+
+mod highlighter {
+    use std::collections::HashMap;
+    use std::collections::HashSet;
+
+    use std::iter::IntoIterator;
+
+    use super::StyleFunc;
+    use crate::model::taggedtext::TaggedText;
+    use crate::model::texttag::*;
+
+    use tui::style::{Color, Style};
+    use tui::widgets::Text;
+
+    pub fn highlight_tagged_text<'a>(
+        text: &'a TaggedText,
+        tag_to_func: &HashMap<Tag, StyleFunc>,
+    ) -> Vec<Text<'a>> {
+        let mut highlighted_text: Vec<Text> = Vec::new();
+
+        // break the boundaries of the Tags into two queues
+        let tags = text.tags();
+        let tagged_str = text.as_str();
+        let mut start_indices = Vec::new();
+        let mut end_indices = Vec::new();
+        for tag in tags {
+            let start_idx = tag.start_idx();
+            let end_idx = tag.end_idx();
+
+            start_indices.push((start_idx, tag.tag()));
+            end_indices.push((end_idx, tag.tag()));
+        }
+
+        let mut active_tags = HashSet::new();
+        let mut last_idx = 0;
+        let mut left = 0;
+        let mut right = 0;
+        while left < start_indices.len() || right < end_indices.len() || last_idx < tagged_str.len()
+        {
+            let mut chunk_str = "";
+
+            if left == start_indices.len() && right == end_indices.len() {
+                chunk_str = &tagged_str[last_idx..tagged_str.len()];
+                let styled_chunk =
+                    Text::styled(chunk_str, style_from_tags(&active_tags, tag_to_func));
+                highlighted_text.push(styled_chunk);
+                break;
+            }
+
+            if last_idx == tagged_str.len() {
+                // Cursor should be the only tag at the end of the string
+                active_tags.clear();
+                active_tags.insert(Tag::Cursor);
+                let styled_chunk =
+                    Text::styled("█", Style::default().bg(Color::Black).fg(Color::White));
```

Comment:
Are the "█" used to highlight the text?

Edit: Ah, correct me if I'm wrong but it seems like this is the cursor that is at the end of a line?

Reply:
yeah it's for when it's at the end of the line. For some odd reason I couldn't get it to work if it was just \n. i had to add a block character before it